linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala

```diff
@@ -3137,20 +3137,69 @@ private class FunctionEmitter private (
   private def genArrayValue(tree: ArrayValue): Type = {
     val ArrayValue(arrayTypeRef, elems) = tree
 
-    val expectedElemType = arrayTypeRef match {
-      case ArrayTypeRef(base: PrimRef, 1) => base.tpe
-      case _                              => AnyType
-    }
-
-    // Mark the position for the header of `genArrayValue`
     markPosition(tree)
 
-    SWasmGen.genArrayValue(fb, arrayTypeRef, elems.size) {
-      // Create the underlying array
-      elems.foreach(genTree(_, expectedElemType))
+    arrayTypeRef match {
+      case ArrayTypeRef(base: PrimRef, 1) if elems.forall(_.isInstanceOf[Literal]) =>
+        // Use a constant array in a data segment
+        val length = elems.size
 
-      // Re-mark the position for the footer of `genArrayValue`
-      markPosition(tree)
+        val (dataID, offset) = (base.charCode: @switch) match {
```

Comment:
Any reason you are not matching on `base.tpe` for better correctness / exhaustiveness checking?

Reply:
Oops, no reason. Fixed. That's better indeed.